enhancements/separating-provider-specific-code-in-the-installer.md

```diff
@@ -0,0 +1,147 @@
+---
+title: separating-provider-specific-code-in-the-installer
+authors:
+  - "@janoszen"
+reviewers:
+  - "@Gal-Zaidman"
+approvers:
+  - TBD
+creation-date: 2020-11-27
+last-updated: 2020-11-27
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Separating provider-specific code in the installer
+
+## Release Signoff Checklist
+
+TBD
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+This change introduces an internal API in the installer code to move provider-specific code into a separate folder and avoid having large switch-case structures in the installer code.
+
+## Motivation
+
+Several parts of the installer code have large, [unwieldy switch-case structures](https://github.com/openshift/installer/blob/254680f6268a7b8e57091d788256c67b81755b5d/pkg/asset/cluster/tfvars.go#L166) that contain provider-specific codes. This makes it hard to keep track of the code that serves a provider, and it makes it impossible to write component-level tests. Creating a separated API reduces friction between teams and makes changes easier to merge because the risk of accidental side effects between providers is reduced. 
+
+### Goals
+
+Create a Go interface and the hooks in the installer code that separatees the installer core code from the provider-specific code.
+
+### Non-Goals
+
+Remove all provider-specific pieces from the code. That would require a complete refactor of the install config data structure and move the Terraform code. 
+
+## Proposal
+
+### Step 1: Create a Go interface
+
+We should create a Go interface that describes the functionality each provider must implement. For example, from the prototype:
+
+```go
+type PlatformV2 interface {
+	// Name returns the name of the platform.
+	Name() string
+
+	// Metadata translates the installConfig to the appropriate entry in clusterMetadata.
+	Metadata(clusterMetadata *types.ClusterMetadata, installConfig *installconfig.InstallConfig) error
+
+    //...
+}
+```
+
+Not all platforms implement all features. Optional features should be hidden behind a feature flag and a separate interface:
```

Comment:
Actually, there is a much better way to do this as I found out. We can create a struct that has receivers for the default functionality and providers can embed this default struct. (Basically, OOP inheritance.)

Reply:
I am not sure I understand. The way I'm figuring this, Installer-core defines a struct with optional methods defined as no-ops, then the Platform embeds it in a struct where it can override the optional noops with actual logic, then the Installer-core imports that... I see a circular dependency

Reply:
In other places where I've created driver-like APIs of this sort, repeating some code always gave us a result that was easier to maintain over time because a developer working on one driver does not have to untangle an inheritance hierarchy or track down the real implementation that may live somewhere else. In places where we could gain from code reuse, utility functions or types that are instantiated rather than inherited are simpler to test and more flexible.

Along the same lines, no-op implementations, even if they have to be provided by every platform, are simpler to understand than features flags and multiple interfaces.